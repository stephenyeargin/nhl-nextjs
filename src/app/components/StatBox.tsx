import React from 'react';
import { formatStat } from '@/app/utils/formatters';

type StatHeader = {
  key: string;
  label: string;
  title: string;
  altKey?: string;
  precision?: number;
  unit?: string;
};

interface StatBoxProps {
  statKey: string;
  value: number | string | undefined;
  statHeaders: readonly StatHeader[];
}

const StatBox: React.FC<StatBoxProps> = ({ statKey, value, statHeaders }) => {
  const header = statHeaders.find((s) => s.key === statKey || s.altKey === statKey);
  const title = header?.title ?? statKey;
  const precision = header?.precision ?? 0;

  const content = (() => {
    if (statKey === 'plusMinus') {
      if (value === undefined || value === null || value === '') {
        return '--';
      }
      const n = typeof value === 'number' ? value : Number(value);
      if (Number.isNaN(n)) {
        return '--';
      }

      return n > 0 ? `+${n}` : `${n}`;
    }

    return formatStat(value, precision);
  })();

  return (
    <div
      className="p-2 bg-transparent text-center border rounded content-center"
      style={{ minWidth: '7rem' }}
    >
      <div className="text-2xl capitalize">{content}</div>
      <div className="text-xs font-light">{title}</div>
    </div>
  );
};

export default StatBox;
