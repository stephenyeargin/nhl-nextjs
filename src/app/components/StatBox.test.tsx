import React from 'react';
import { render, screen } from '@testing-library/react';
import StatBox from '@/app/components/StatBox';

const headers = [
  { key: 'plusMinus', label: '+/-', title: 'Plus/Minus' },
  { key: 'goals', label: 'G', title: 'Goals' },
] as const;

describe('StatBox', () => {
  it('renders + sign for positive plusMinus', () => {
    render(<StatBox statKey="plusMinus" value={3} statHeaders={headers} />);
    expect(screen.getByText('+3')).toBeInTheDocument();
  });

  it('renders negative value for minus plusMinus', () => {
    render(<StatBox statKey="plusMinus" value={-2} statHeaders={headers} />);
    expect(screen.getByText('-2')).toBeInTheDocument();
  });

  it('renders -- for undefined plusMinus', () => {
    render(<StatBox statKey="plusMinus" value={undefined} statHeaders={headers} />);
    expect(screen.getByText('--')).toBeInTheDocument();
  });

  it('renders plain number for regular stat', () => {
    render(<StatBox statKey="goals" value={10} statHeaders={headers} />);
    expect(screen.getByText('10')).toBeInTheDocument();
  });

  it('matches snapshot (plusMinus positive)', () => {
    const { container } = render(<StatBox statKey="plusMinus" value={4} statHeaders={headers} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
