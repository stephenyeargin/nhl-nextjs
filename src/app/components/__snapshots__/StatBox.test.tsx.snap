// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatBox matches snapshot (plusMinus positive) 1`] = `
<div
  class="p-2 bg-transparent text-center border rounded content-center"
  style="min-width: 7rem;"
>
  <div
    class="text-2xl capitalize"
  >
    +4
  </div>
  <div
    class="text-xs font-light"
  >
    Plus/Minus
  </div>
</div>
`;
